ic-helpers/src/factory/api.rs

```diff
@@ -210,3 +214,217 @@ macro_rules! init_factory_api {
         }
     };
 }
+
+/// This macro is essentially the same as [`init_factory_api`], but it is used
+/// to extend the canister api, generated from the `Canister` macro in `ic-canister-macros`.
+///
+/// API methods that are added:
+/// * get_checksum
+/// * get_cycles
+/// * top_up
+/// * upgrade
+/// * version
+/// * length
+/// * get_all
+/// * get_icp_fee
+/// * set_icp_fee
+/// * get_icp_to
+/// * set_icp_to
+/// * get_controller
+/// * set_controller
+/// * refund_icp
+#[macro_export]
+macro_rules! extend_with_factory_api {
+    ( $canister:ident, $upgrading_bytecode:expr ) => {
+        impl $canister {
+            /// Returns the checksum of a wasm module in hex representation.
+            #[query]
+            #[candid_method(query)]
+            fn get_checksum(&self) -> String {
+                self.state.borrow().factory().checksum.to_string()
```

Comment:
The state parameter is not necesserilly called `state` in the canister. It can be any other name. So it's better to have this name as the parameter of the macro, and write `self. $state` here intead.